#1201 - Get health facilities with google places api

maskowe opened the pull request:
### What does this pull request aim to achieve? 💯
In backend, provides data necessary for client to display list and map for health facilities
_Please be concise and link any related issue(s):_
For the backend, this PR resolves #1201 
<!--### 🚨Before submitting this pull request🚨:-->

<!--_Please do **NOT** submit this PR if you have not done the following:_-->

- [X] I have checked that no one else is working on similar changes.
- [X] I have read the latest [**README**](README.md) and understand the development workflow.
- [X] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [X] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [X] There are no merge conflicts.
- [X] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [X] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <#1201 Get health facilities data>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

Commit:
1201 initial commit

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

ddveloper commented on backend/lib/endpoints/schema/geo.js:
```diff
@@ -21,8 +21,15 @@ const getLocationReverseGeocodeSchema = {
 
+const getHealthFacilityPlacesSchema = {
+  querystring: strictSchema()
+    .prop("lat", S.number().minimum(0).maximum(90).required())
```
I am not sure but should this be minimum(-90)?

maskowe replied:
Copied validity requirements from getLocationReverseGeocodeSchema.   Also  googled to verify and got this:  Latitude : max/min +90 to -90
Longitude : max/min +180 to -180.
But google maps throws error when lat is less than zero (I tested this) so I think that getLocationReverseGeocodeSchema requirements are probably right.   Josh M - Can you verify?  I can do further research if necessary.

idjevm replied:
@joshmorel thoughts here?

joshmorel replied:
> But google maps throws error when lat is less than zero

That would be kind of limiting...

https://github.com/googlemaps/google-maps-services-js/blob/542eb964317f79fd7aa82d4a39d2df834af8070e/src/common.ts#L480

@maskowe  They accept different formats types for location (string, array, object). I'd use the same object format as I did for reverseGeocode.latlng as it's explicit which is lat & which is lng (not an array), e.g.:

```
        location: {
          lat: 51.580559,
          lng: -0.3419949999999999,
        },
```

idjevm commented on backend/lib/components/Geo/gmaps/getDocFacilities.js:
```diff
@@ -0,0 +1,20 @@
+    config: { geo },
+  } = require("../../../../config");
+  
+var places = [];
+
+const getDoctorFacilities = async (location, maxFacilities) => { 
+    const client = new Client({});
+    await client.placesNearby({ params: { location: location, type: 'doctor', rankby: 'distance', key: geo.googleMapsApiKey, }
+    })
```
```suggestion
const getDoctorFacilities = async (location, maxFacilities) => { 
    let places = [];
    const client = new Client({});
    await client.placesNearby({ params: { location: location, type: 'doctor', rankby: 'distance', key: geo.googleMapsApiKey, }
    })
```

maskowe replied:
Would rather not use let since the block script could be too strict in this case.   Would like to leave this as var. Here's a good discussion.   https://stackoverflow.com/questions/762011/whats-the-difference-between-using-let-and-var
Look at "Scoping Rules" down the page that got 6307 upvotes.  
Probably in this case, let would be too strict for scoping.

idjevm replied:
Ohh it should be fine, the variable places is not being used anywhere else so this suggestion should work fine. And we need to avoid as much as possible using `var`...

maskowe replied:
Ok - I see what you mean - I was afraid of underscoping, and you would rather not overscope...So I'm changing all the vars to lets.

joshmorel replied:
@maskowe I wouldn't mix the use of await & then/catch here. I'd try to be consistent with async/await like the rest of the app - for consistency at least but it's also more readable. This could be re-written (for example):

```javascript
const { data }  = await client.placesNearby({ params: { location, type: 'doctor', rankby: 'distance', key: geo.googleMapsApiKey });
return data.results.length > 0 ? data.results.slice(0, maxFacilities) data.results;
}
```

joshmorel replied:
But let's wait til we resolve the question about which type to use. I don't think we'll need to this file in any case. 

I do think there will be a few other eslint issues here. Unfortunately it's not working with docker right now (open PR I need to review #1600).

idjevm commented on backend/lib/components/Geo/gmaps/getFacDetails.js:
```diff
@@ -0,0 +1,50 @@
+    "opening_hours"
+  ];
+
+var placesWithDetails = [];
```
Same as above, is there any reason why defining here?

maskowe replied:
Constants are better off sharable for any new exported functions, I think.   I did move the array variables inside the functions, though.   thanks.

idjevm replied:
These are not using anywhere else, let's move this variable inside the method and have it return it in case it is needed anywhere else. 

Also we need to make sure we are not using `var`.

maskowe replied:
Ok - changing all vars to lets here also.

idjevm commented on backend/lib/components/Geo/gmaps/getFacDistances.js:
```diff
@@ -0,0 +1,41 @@
+    config: { geo },
+  } = require("../../../../config");
+
+var placesWithDistances = [];
+
+const getFacilityDistances = async (location, places) => { 
```
Same as above.

maskowe replied:
Moving array variable inside.

idjevm replied:
Also, we should change any `var` to `let`, and make sure they are in the right scope.

maskowe replied:
Ok - changed to let.

idjevm commented on backend/lib/components/Geo/gmaps/index.js:
```diff
@@ -50,8 +54,49 @@ const getLocationByLatLng = async (lat, lng) => {
+  const MAX_DOCTOR_FACILITIES = 5;
+  const location = [lat, lng];
+  
+  var places = [];
```
Same as above.

maskowe replied:
I think that the constants here are special purpose for the health facilities only, as is the places array.   Would like to leave these here.

maskowe replied:
Would also like to leave the **require** constants outside, as they could potentially be reused for the module in other functions/endpoints.

idjevm replied:
yeah definitely leave that const to require outside, but let's definitely move the `places` array inside the method, also change any `var` to `let`. If we needed the variable `places` then we could use the method to get that.

maskowe replied:
Ok - done

idjevm reviewed (changes requested):
Let's use `let` instead of `var`. I left some other comments so far, I will dig more into it later...

idjevm reviewed (changes requested):
I think we should def change the `var` to `let` since we would want to keep those scoped, also, if we do need that variable somewhere else, we would want to use the function to get it as the variable alone is just an empty array...

joshmorel reviewed (changes requested):
Thanks @maskowe !

My big first question for this was around why there is a separate query/function for hospitals & doctors. I did some digging and see they only accept 1 type which explains it. I would use a single function with different parameters if we need to but I'm questing using the `'doctor'` type.

A "place" in Google Maps have have more than 1 type. It seems all these doctor ones also have the `'health'` type as do those with `'hospital'`. So we could just use `'health'` however not sure if that's the best approach?

`'health'` may be a little too broad. For one thing it includes all these doctors... In my country if you give a visitor seeking medical attention a list littered with private doctor's offices (that aren't also walk in clinics) it would be considered bad advice. If you aren't on their "roster" (already a patient) they can't help you, they'll tell you to call 911 or find another place that accepts walk in patients. Pharmacies are actually more useful in such a case if you have a non-urgent health question.

Not sure what it's like in other countries... but it seems `'hospital'` might filter out that "noise". It does seem to include walk in clinics that aren't exactly capital "H" hospitals. 

See: https://maps.googleapis.com/maps/api/place/nearbysearch/json?location=43.472165,-80.538425&radius=1500&type=hospital&key=YOURKEY

```
  {
         "business_status" : "OPERATIONAL",
         "geometry" : {
            "location" : {
               "lat" : 43.4717593,
               "lng" : -80.5387664
            },
            "viewport" : {
               "northeast" : {
                  "lat" : 43.4731694802915,
                  "lng" : -80.5373502697085
               },
               "southwest" : {
                  "lat" : 43.4704715197085,
                  "lng" : -80.54004823029152
               }
            }
         },
         "icon" : "https://maps.gstatic.com/mapfiles/place_api/icons/v1/png_71/generic_business-71.png",
         "name" : "Waterloo Walk-In Clinic",
         "opening_hours" : {
            "open_now" : true
         },
         "photos" : [
            {
               "height" : 3400,
               "html_attributions" : [
                  "\u003ca href=\"https://maps.google.com/maps/contrib/110429908807247242474\"\u003eWaterloo Walk-In Clinic\u003c/a\u003e"
               ],
               "photo_reference" : "CmRaAAAArSBWRLr-Zn3dEiIBoQ8iLpHwYAwfAbziweZ2A6v0gWeS3QZx_6LV9dquX6TpDSPVzuZSzbtpsb7CVxsOq2ywDo2bqjjaIk6ANcRHz8X2hEiZTqav1ry0HuoOLwqTzLWDEhApLeZtKixHTzReuRf31zuHGhSz0xHHdOeccfWolbJ2X-6eCLmEtw",
               "width" : 4449
            }
         ],
         "place_id" : "ChIJQdQ9MgH0K4gRIfW8ouHsR0w",
         "plus_code" : {
            "compound_code" : "FFC6+PF Waterloo, ON, Canada",
            "global_code" : "86MXFFC6+PF"
         },
         "rating" : 3.6,
         "reference" : "ChIJQdQ9MgH0K4gRIfW8ouHsR0w",
         "scope" : "GOOGLE",
         "types" : [
            "hospital",
            "doctor",
            "local_government_office",
            "health",
            "point_of_interest",
            "establishment"
         ],
         "user_ratings_total" : 745,
         "vicinity" : "170 University Avenue West, Waterloo"
```



Thoughts @davidoduk ? No matter what's better we only need to make one call - `type: "hospital"` or `type: "health"`

rabbi08 commented:
@maskowe @davidoduk 
Tested on Laptop and iphone:

- Unable to see the nearest health facilities to my current location.
- User is directed to 'Feeds' page while answering the onboarding question not list of hospitals as mention in the requirements on the notion page.
Please refer below GIF.

![ezgif com-gif-maker (35)](https://user-images.githubusercontent.com/67168400/97749208-0b931480-1ac5-11eb-8e27-e5ae0576dec9.gif)

Please, let me know if there is some issue or I am missing something.

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
add new type parm and corr lint issues

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

joshmorel commented on backend/lib/components/Geo/gmaps/index.js:
```diff
@@ -50,8 +128,39 @@ const getLocationByLatLng = async (lat, lng) => {
+    .then((result) => {
+      result.forEach(function(item) {
+        const place = new Place(item.name, item.place_id, type); 
+        places.push(place);
```
@maskowe do we need this `Place` class? Looking at the rest of the code it doesn't seem like this functionality is not really used. 

The only difference I see in the output after removing this is lack of `type` with the response, which could be achieved in a much more simple way.

maskowe replied:
Thanks.  I included the type originally in the class because originally the client was not sending that in.   Even so, I think it's more convenient, but if you don't like it, I can certainly take it out.  Also the class includes distance, which extends the rest of the place details, being part of the different api.   I have pretty much built into the place class, since I'm doing the FE stuff and will know the structure based on this class.   But if everything comes back that I need (everything in the place instance), then I'm good with however you want to work that.

maskowe replied:
Some of the functionality is not used (getters and setters), because I thought maybe unit testing could use those...Again, can take it out if you want.  It's ok for you to work it however you want, just let me know and thanks again.

joshmorel replied:
>  Also the class includes distance, which extends the rest of the place details, being part of the different api.

Aren't you getting distance here? 

https://github.com/FightPandemics/FightPandemics/blob/b5bf662f4484a007e8589b02ad8a23410345a5cd/backend/lib/components/Geo/gmaps/index.js#L71

maskowe replied:
Yes - distance: data.rows[0].elements[i].distance.text  provides the distance by index position to the place instance property called "distance".   That's what I meant when I said "extends" the class..

joshmorel reviewed (dismissed):
@maskowe I'm working on applying linting and can also address the issue with an alternate solution if you like. I just want to make sure I'm not missing something...

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

joshmorel committed:
refactor geo facilities backend & fix 2 issues (#1883)

issue:
-accept to -90 latitude
-handle no distancematrix result

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

joshmorel commented:
Will re-review with FE additions

beth-maskow committed:
Merge branch 'feature/1201-display-health-facilities' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
FE changes with new and updated components

beth-maskow committed:
removed console logs

ddveloper commented on client/src/components/NearestHospital/FacilityCard.js:
```diff
@@ -0,0 +1,177 @@
+
+  const getOpenDays = (periods) => {
+    if (periods.length === 0) return "";
+    if (
```
I don't understand this part, maybe adding some comments here?

maskowe replied:
I will do that.   The backend google api will not send periods if that information wasn't provided, so it results in empty space.

maskowe replied:
done

ddveloper commented on client/src/components/NearestHospital/CheckSymptomsBox.js:
```diff
@@ -45,14 +43,26 @@ const SymptomsLink = styled(NavLink)`
-        <SymptomsLink to="/symptoms-check">Check Symptoms</SymptomsLink>
+      <div>
+        <p>
+          <span>{t("nearestHsp.hlthEval")}</span>
```
hlth to health?

maskowe replied:
This component will be developed in the future with another PR/feature branch.   But I will update it.

maskowe replied:
done

ddveloper commented on client/src/components/NearestHospital/CheckSymptomsBox.js:
```diff
@@ -45,14 +43,26 @@ const SymptomsLink = styled(NavLink)`
+          to="/symptoms-check"
+          id={GTM.nav.prefix + GTM.nav.checkSymp}
+        >
+          {t("nearestHsp.chkSymptms")}
```
symptms to symptoms?

maskowe replied:
done

ddveloper commented on client/src/components/NearestHospital/LocalEmergencyNumber.js:
```diff
@@ -67,9 +68,11 @@ const EmergencyLine = styled(NoticeBar)`
   return (
     <EmergencyLine mode="closable" icon={null} {...props}>
-      <h4>Local emergency number</h4>
+      <h4>{t("nearestHsp.locaEmrgNbr")}</h4>
```
locaEmrgNbr to localEmergencyNumber?

maskowe replied:
This component will be developed in the future with another PR/feature branch.  (probably by me in the near future).   But I will update it with a new commit.

maskowe replied:
done

ddveloper commented on client/.env:
```diff
@@ -1 +1,2 @@
 EXTEND_ESLINT=true
+REACT_APP_GOOGLE_KEY=AIzaSyAeoBJPdX8fHzuIxXA6vIknwyceEuut3tY
```
I don't know if there is a better way to use the google key, @Zaydme  any ideas?

maskowe replied:
@ddveloper @Zaydme I've left a DM for Zaydme in regards to this.   Right now, I'm writing an api to encrpt the key from the server, and remove this env variable from the FE so that the client can decrypt it when it uses the key.   (using CryptoJS).    Since this exposes the key to devtools even if the env file is hidden.   Also I have Dawei Zhang researching for verification on this solution.   Will commit these changes when I get this portion done on the backend.

ddveloper replied:
add @mannykary

maskowe replied:
It looks like we need a "browser" key for this, (keeping the backend key the same).   So I think we need two different keys (the browser key being on the FE.

https://webmasters.stackexchange.com/questions/95918/restricting-google-maps-js-api-keys-to-just-my-domain

I think we need to ask Josh Morel or (Manny possibly) to create a new key and add the domain restriction(s).  Until they respond, I think we're ok leaving the FE env variable (key) in place for just the PR to test.

joshmorel replied:
@maskowe you can and we should restrict based on referrer domain which is set in Google Maps Cloud console. The key shouldn't be in source control anyways but you dont need to encrypt it.. (due to referrer limitation). Check out what is done with GTM key (cc @mannykary )

mannykary replied:
@maskowe See https://github.com/FightPandemics/FightPandemics/blob/staging/.github/workflows/production-workflow.yml#L26-L29. For the GTM key, we added it to Github Secrets and then loaded it in when the client Docker image is built. We will need a Google Key for each environment (Review, Staging, and Production). You can DM me or anyone with admin repo access (I believe @joshmorel, @ddveloper, and @Zaydme have access) so that the keys can be added to Github Secrets.

maskowe replied:
@mannykary Thanks - Josh offered to help soon.   If he's unable for time restrictions, I'll let you know.   I see how to refer to this, 

client/src/index.js
  gtmId: process.env.REACT_APP_GTM_ID,
};

maskowe replied:
@mannykary Josh M. has added the three browser keys we need as:

created the 3 keys in google cloud console:
FE Maps API Key - Staging: please see private slack message
FE Maps API Key - Review: please see private slack message
FE Maps API Key - Production: please see private slack message

Can you possibly get the time to add these to github secrets?

He has added the keys as values to corresponding GitHub secrets keys:
STAGING_FE_GOOGLE_MAPS_API_KEY
REVIEW_FE_GOOGLE_MAPS_API_KEY
PRODUCTION_FE_GOOGLE_MAPS_API_KEY

And I will do the yml changes needed.   
Thanks again.

maskowe replied:
Or you can DM me the link and credentials with instructions so that I can try to do this myself, if time is an issue.

maskowe replied:
Keys have been addressed for all branches.   Please retest at https://feature1201-display-health-facilities.fightpandemics.xyz

ddveloper commented on backend/lib/components/Geo/gmaps/index.js:
```diff
@@ -50,8 +66,85 @@ const getLocationByLatLng = async (lat, lng) => {
+    open_now:
+      data.result.opening_hours === undefined || data.result.opening_hours.open_now === undefined
+        ? ""
+        : JSON.stringify(data.result.opening_hours.open_now),
```
maybe leave to front-end to handle the formatting?

maskowe replied:
Thanks, ddveloper.  Would like to leave this if it's ok, since it's easier to not have to have the FE parse this out.

ddveloper commented on backend/lib/endpoints/schema/geo.js:
```diff
@@ -21,8 +27,22 @@ const getLocationReverseGeocodeSchema = {
+        .enum(Object.values(HEALTH_FACILITY_TYPES))
+        .default(HEALTH_FACILITY_TYPES.HOSPITAL),
+    ),
+};
```
if the response format is finalized, we may add the response schema inside, e.g. 
```
  response: {
    200: S.object()
      .prop(...
  },
```
more examples in this [link](https://github.com/FightPandemics/FightPandemics/pull/2049/files)

maskowe replied:
@joshmorel may have thoughts on this, since he's already code-reviewed this and passed it.  Thanks.

ddveloper reviewed (commented):
There might be changes requirement (like scrolling doesn't affect the map, layout, etc.) from design team, @CamilaArias @rabbi08 ?

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
yml updates, fix scroll map and others

beth-maskow committed:
remove review key entry from env 01122021

maskowe commented:
> @maskowe @davidoduk
> Tested on Laptop and iphone:
> 
> * Unable to see the nearest health facilities to my current location.
> * User is directed to 'Feeds' page while answering the onboarding question not list of hospitals as mention in the requirements on the notion page.
>   Please refer below GIF.
> 
> ![ezgif com-gif-maker (35)](https://user-images.githubusercontent.com/67168400/97749208-0b931480-1ac5-11eb-8e27-e5ae0576dec9.gif)
> 
> Please, let me know if there is some issue or I am missing something.

Please retest at https://feature1201-display-health-facilities.fightpandemics.xyz
If possible.

rabbi08 reviewed (changes requested):
Thanks @maskowe , Found below observations while testing:

1. The Results for the Pharmacies also display the companies at the top. See below:
- Search 'Plainsboro Township'
  Ranbaxy Pharmaceuticals Inc
- Search 'Princeton, NJ' 
   Proximare Inc: Proximare is an advisory firm

![ezgif com-video-to-gif-4](https://user-images.githubusercontent.com/67168400/104856043-30d23a80-58de-11eb-9d61-6ca93b8f4ed4.gif)

Please see below screen recording with google maps result set:

2. The search results should be sorted as per location and category matched.
As, Currently, locations farther are displayed before other nearest ones. Please in screenshot below.

<img width="1021" alt="image" src="https://user-images.githubusercontent.com/67168400/104854849-04ff8680-58d7-11eb-9185-057b27c2db83.png">

3. The icon's description on the maps are overlapped and the tooltips on the icons are not working properly, when overlapped. Please see below screenshot and screen recording.

<img width="781" alt="image" src="https://user-images.githubusercontent.com/67168400/104854877-35dfbb80-58d7-11eb-9d20-bf7ffa0e155b.png">

![ezgif com-video-to-gif-3](https://user-images.githubusercontent.com/67168400/104854997-0d0bf600-58d8-11eb-9e5a-93984e2c8beb.gif)

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
correct sorting, invalid places and marker overlap

maskowe commented:
@rabbi08 Thanks a lot for the thorough review.   In regards to:

1)  With regards to Plainsboro Township, have eliminated Ranbaxy Pharmaceuticals Inc  and others that Google maps nearby search doesn't include.   Evidently, users have incorrectly identified with "pharmacy" as their category, even though they are not "pharmacies" but other types of establishments.
       
  I have also included a new link under the type header "For more results for this facility type and location" which should fill in the gaps with the relevant facilities.   Google maps nearby search leans heavily on reviews and other metadata to determine what to show and not show.   For pharmacies, we should be cautious with our exceptions.   My solution is applicable to all locations for the pharmacy types.

For Proximare Inc -  Princeton, NJ
  Google identifies this as a pharmacy also, so it must have eliminated it from the nearby list based on some other metadata that it has (or doesn't have).   We don't have access to this metadata and can't eliminate based on the generic name, so we have to include it as long as it's within the proximity set of results.

![proximare google image](https://user-images.githubusercontent.com/46111253/105389370-240b5a80-5bee-11eb-8a4b-2695b5488f8d.png)

2) Even though sorting by distance wasn't a requirement initially, I've added this feature.

3)  I've eliminated the "overlap" of the markers into the description boxes when hovering over markers.  Any other issues with those markers that are overlapping each other should be easily manually worked around by expanding the map itself.   If you find that this doesn't address your issues, please let me know.

Thanks again - ready to retest.

ddveloper commented:
@CamilaArias would you also see if any comments from design perspective? thanks 🎨

rabbi08 commented:
@maskowe @davidoduk 
I have following observations:
1. The distance is displayed in 'km' but in countries like 'US' distance unit is 'Miles'. @doheekim93 suggested to add it to Lokalise for customization,  so I will add new ticket for the same.

2. The results doesn't seem to be accurate as in the search. While searching for 'Pharmacy for zip code = 08536'
There are 2 pharmacy near to each other with almost same address but one is displayed and other is not. Please see below GIF: where nearest CVS location is not displayed:

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/67168400/105896554-f30f9900-5fe4-11eb-9c55-174bc315a3b2.gif)

3. We need to have GTM IDs for "Health Aid, location input, Hospitals, Pharmacies, Doctors" so total 5 GTM IDs.

Let me know if there are any questions/comments. Thanks

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
include mi for distance units and add GTM id

maskowe commented:
@rabbi08 Thanks again.  
1)  Distance now defaults to miles (as in "6 mi") in the description for those (3) countries using imperial instead of metric system.
2)  The CVS drug store very close to Plainsboro pharmacy is identified as a "drug store" instead of a "pharmacy" when I search google maps nearby.   My guess is that since the business user that entered this information left off "pharmacy" and the description shows only "beauty supplies and home health items", it was not included in our results for "pharmacy" type.   This is more bad user data, that's why I've included the new link "For more results" in order for the FP users to make their own selection from each set.   Google has the metadata to make intelligent guesses about these pharmacies that we just can't get to with this API, unfortunately.   
![cvs drug store google map image](https://user-images.githubusercontent.com/46111253/106156118-8a94f900-6157-11eb-978f-07b9b8e5b3f2.png)
3)  Actually, there are six elements, since there are two instances of LocationInput (one for the "Share your location..." element and another for the change location element).
For facility type buttons ("Hospitals, etc."):  id/selector: a#NH_FT0, a#NH_FT1, a#NH_FT2
For the two location input (text boxes): id/selector: span#NH_LC_EAZ and span#NH_LS_EAZ
 For the Health Aid link (top nav):  id/selector:  a#NAV_NH
I did have to add a property to the tags in constants file for the buttons, so thanks for the catch.

rabbi08 commented:
@maskowe 
I have following observations:
- Please add scroll bar in the search results, when user scroll down in search results, only the search results should scroll and map should stay as is in the right pane of the window.
- Also, for Point#2< "user bad data" for 'Pharmacy'> agree with your comments, but while searching for hospitals, also I am not getting expected results. Great if you can look into this as well. Please see below GIF for search result comparison:

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/67168400/106846618-391acb80-667b-11eb-8a3a-e8b1fa0a9d2d.gif)

- The search result display two hospitals with same location but different name( one is hospital and other is medical facility/center) Please see below screenshot

<img width="1440" alt="Screen Shot 2021-02-03 at 11 59 27 PM" src="https://user-images.githubusercontent.com/67168400/106847055-2b197a80-667c-11eb-9963-a6aff1d3a29c.png">
 
- the text, to show more search results, please update the text:
Update: "For more results for this facility type and location"
To: "Show more search results..."
Currently, this text navigate user to google maps. Shouldn't we keep user on our portal rather than redirecting to google. Great if we add pagination to show results in FightPandemics only, rather than redirecting, if that's feasible. 

**Please confirm this scenario @davidoduk @maskowe** 

<img width="1440" alt="Screen Shot 2021-02-04 at 12 02 52 AM" src="https://user-images.githubusercontent.com/67168400/106847561-243f3780-667d-11eb-95de-8aaa8109a6ce.png">

ddveloper reviewed (dismissed):
LGTM, thanks @maskowe

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
show res url chgs and fac list scroll added

maskowe commented:
@rabbi08 Thanks again.   I've moved the url for showing more results and changed text as requested.   Also added scrolling for the facility list on left, keeping map in position.  
On the issue with Fresenius Medical Care in Trenton, NJ, I believe that this is Google metadata putting this in the hospital list, even though the establishment category is medical center.   It could be that these "medical centers" are included, we just haven't identified them in the hospital lists coming back from our api.   The Fresenius facility does show up at http://nj.hospitals-usa.org/item/fresenius-medical-care-mercer-capital-health-system/
so maybe that is why.
![fresenius med care nj](https://user-images.githubusercontent.com/46111253/106914407-5c756300-66db-11eb-9038-d1608fbb5b13.PNG)

maskowe commented:
@rabbi08 Also on the pagination topic, we have purposely limited the output to 10 facilities due to bad performance beyond.   We can theoretically make another request for the next 10 (up to 20 only unfortunately).   But since google maps provides unlimited amount and have included/disincluded various facilities because of different factors, I don't believe that this is worth the effort and would also be misleading because of the unfortunate mismatch of both the results themselves and the presentation differences.   We can talk more in slack on this if you want.

rabbi08 commented:
Thanks @maskowe 
1. Ok I agree with your comments for search results difference from google.
2. For scroll bar, please add the scroll bar to the search results and only search results should scroll not the header like 'Hospitals/Doctors/Pharmacy'. 
3. Also, the the length of the search result pane should match with Google maps.

Please see below GIF.

![ezgif com-video-to-gif-3](https://user-images.githubusercontent.com/67168400/107302062-b616be00-6a4a-11eb-811f-5d4f37584a74.gif)

rabbi08 commented:
@maskowe 
Following are the observations while testing this feature on mobile:

1. For some options on **both mobile and Desktop** it is displayed as 'common.myqrganisation', 'common.language', 'common.no' there should not be 'common.<option>' . Please see below screenshot and in below GIF in #4 too. 
<img width="1440" alt="Screen Shot 2021-02-09 at 12 25 28 AM" src="https://user-images.githubusercontent.com/67168400/107319818-904ee080-6a6d-11eb-81a9-0a707ecb6515.png">

2. 'Map view' text is displayed over the text in the search results. Please see below screenshot.
![IMG_4808](https://user-images.githubusercontent.com/67168400/107319534-01da5f00-6a6d-11eb-9fc3-0e7dbadee42f.PNG)

3. The scrolling is not smooth i.e. initially the search options are not displayed when user enter the location and navigate to the search results page, the 'Facility options-Hospitals,,, etc. " are NOT displayed. Please see below screenshot 
![IMG_4808](https://user-images.githubusercontent.com/67168400/107319489-ef602580-6a6c-11eb-809d-eec5cffd061e.PNG)

4. The user need to scroll 2 times to view the Facilities. Please see below GIF showing the navigation.
![ezgif com-video-to-gif-4](https://user-images.githubusercontent.com/67168400/107319621-29312c00-6a6d-11eb-89f3-9169dc5f6797.gif)

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
disallow scrl on fac hdr and adj map vw btn

beth-maskow committed:
smooth scrolling

maskowe commented:
Changed scrolling list to disinclude header, smoothed scrolling behavior and adjusted position of Map View/List View button to allow for some mobile devices to allow it to be shown.   Lambda testing showed that iPhones (most or all versions) not only do not show scroll bar for list, but actually introduce an upper scroll bar that other devices don't shown.   Needs to be tested outside of Lambdatest (on a real phone) to confirm this.

rabbi08 commented:
Thanks @maskowe 

On the Desktop, the scrolling and width issues are working fine, Hence no issues have been observed in the desktop version.

But On MOBILE, (iPhone 11 Pro max) the 'Map view' is getting scrolled up... As per design, it should display in the Bottom of the page and centre aligned and should not be scrolled while user scroll the result sets. 
Please see below GIF from mobile version:
![ezgif com-video-to-gif-8](https://user-images.githubusercontent.com/67168400/107983734-6c6f2b80-6f94-11eb-9d42-1b019f1503f1.gif)

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
attempt scrollto top for iphone

beth-maskow committed:
changed styling for iphone issues

beth-maskow committed:
adjust psn of map vw btn top elem

beth-maskow committed:
another adj for map vw btn and top ctnr

maskowe commented:
Ready to retest iphones.   Because some older phones were hiding part of the view map button, I had to move it up a little.   So it appears higher on other phones.   Also, in order for the button to be properly positioned, had to fix the position of the facility type buttons and location and move the facility type header down, so there is some white space, but it had to happen to accommodate iphones.   Thanks.

rabbi08 commented:
Thanks @maskowe 

While testing on mobile view (iPhone 11 Pro max):
1. The 'Map view' option is displayed in the center of the page, it should be in the Bottom of the page century aligned and should be in one line.
2. There is white space over the list like in below GIF, notice the white space over the 'Hospitals' section.
3. When user scrolls within the list the white space exists and when user scroll on the white space 'White space' is gone and issue is fixed.

For the issues above please refer below GIF.
![ezgif com-video-to-gif-9](https://user-images.githubusercontent.com/67168400/108539267-c4b46f00-72ad-11eb-94bd-b9f38111cbd3.gif)

rabbi08 commented:
adding more observation from Mobile view (iPhone 11 Pro max):

1. On Health Aid page -> open the navigation menu from the top left menu bars, you will observe that, the navigation bar is overlapped by the Health Aid options. Please see below in below screenshot and video:
![IMG_279B06AE1094-1](https://user-images.githubusercontent.com/67168400/108540043-b7e44b00-72ae-11eb-837a-0d7788649802.jpeg)


2. The 'languages' option is still displayed as 'Common.languages' hence the prefix 'common.' exist with the option of the menu bar and pop ups.

https://user-images.githubusercontent.com/67168400/108539761-65a32a00-72ae-11eb-987a-6221f71c5dbb.MP4

beth-maskow committed:
fixed top overlp issue and whitespace

beth-maskow committed:
adjust iphone posn down a little

beth-maskow committed:
bumped down top posn again for iphones

beth-maskow committed:
bmpd iphones a little more again

beth-maskow committed:
put back in missing common translations

maskowe commented:
All issue fixed including "Language" bug.
![image](https://user-images.githubusercontent.com/46111253/108562053-a5313c80-72d5-11eb-9271-31b6e5466c5a.png)

kamarenmccord committed:
added optional chaining to facilities on line 122 to prevent fatal error

kamarenmccord committed:
fixed broken css rule on line 20

beth-maskow committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
correct z-index of type and location container, etc.

beth-maskow committed:
Merge branch 'feature/1201-display-health-facilities' of https://github.com/FightPandemics/FightPandemics into feature/1201-display-health-facilities

beth-maskow committed:
move map vw btn down and reduce whitsp for hdr

beth-maskow committed:
adj header down a bit more for iphones

beth-maskow committed:
adjust down again a bit more

beth-maskow committed:
bump down a little more 156pm

beth-maskow committed:
bump again tiny bit more 206pm

rabbi08 reviewed (approved):
thanks @maskowe 
This looks fine on iPhone now. we can merge and test more in STG env  now. Thanks.

@ddveloper @aditilamkhede  great if you can review and merge accordingly. Thanks

idjevm reviewed (approved):
LGTM! There are a few `var` or numbers that could become constants but not blockers.

ddveloper reviewed (approved):
LGTM, thanks @maskowe 👍
